allennlp/models/language_model.py

```diff
@@ -0,0 +1,343 @@
+    to create logits over possible words and then returns the
+    negative log likelihood.
+    """
+    def __init__(self,
+                 num_words: int,
+                 embedding_dim: int) -> None:
+        super().__init__()
+
+        # TODO(joelgrus): implement tie_embeddings (maybe)
+        self.tie_embeddings = False
+
+        self.softmax_w = torch.nn.Parameter(
+                torch.randn(embedding_dim, num_words) / np.sqrt(embedding_dim)
+        )
+        self.softmax_b = torch.nn.Parameter(torch.zeros(num_words))
+
+    def forward(self, embeddings: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        # embeddings is size (n, embedding_dim)
+        # targets is (batch_size, ) with the correct class id
+        # Does not do any count normalization / divide by batch size
+        probs = torch.nn.functional.log_softmax(
+                torch.matmul(embeddings, self.softmax_w) + self.softmax_b,
+                dim=-1
+        )
+
+        return torch.nn.functional.nll_loss(probs, targets.long(), reduction="sum")
+
+
+@Model.register('language_model')
+class LanguageModel(Model):
+    """
+    The ``LanguageModel`` applies a "contextualizing"
+    ``Seq2SeqEncoder`` to uncontextualized embeddings, using a ``SoftmaxLoss``
+    module (defined above) to compute the language modeling loss.
+
+    If bidirectional is True,  the language model is trained to predict the next and
+    previous tokens for each token in the input. In this case, the contextualizer must
+    be bidirectional. If bidirectional is False, the language model is trained to only
+    predict the next token for each token in the input; the contextualizer should also
+    be unidirectional.
+
+    If your language model is bidirectional, it is IMPORTANT that your bidirectional
+    ``Seq2SeqEncoder`` contextualizer does not do any "peeking ahead". That is, for its
+    forward direction it should only consider embeddings at previous timesteps, and for
+    its backward direction only embeddings at subsequent timesteps. Similarly, if your
+    language model is unidirectional, the unidirectional contextualizer should only
+    consider embeddings at previous timesteps. If this condition is not met, your
+    language model is cheating.
+
+    Parameters
+    ----------
+    vocab: ``Vocabulary``
+    text_field_embedder: ``TextFieldEmbedder``
+        Used to embed the indexed tokens we get in ``forward``.
+    contextualizer: ``Seq2SeqEncoder``
+        Used to "contextualize" the embeddings. As described above,
+        this encoder must not cheat by peeking ahead.
+    dropout: ``float``, optional (default: None)
+        If specified, dropout is applied to the contextualized embeddings before computation of
+        the softmax. The contextualized embeddings themselves are returned without dropout.
+    loss_scale: ``Union[float, str]``, optional (default: 1.0)
```

Comment:
We can remove this argument, I don't think it's useful.

Reply:
same as above